src/Workspaces/Remote/Core/RemoteHostAssetWriter.cs

```diff
@@ -0,0 +1,241 @@
+/// its buffer.  Once present in the reader's buffer, synchronous deserialization can happen without any sync-over-async
+/// blocking on async-io.
+/// <para/> The sentinel byte serves to let us detect immediately on the reading side if something has gone wrong with
+/// this system.
+/// <para/> In order to be able to write out the data-length, the writer will first synchronously write the asset to an
+/// in-memory buffer, then write that buffer's length to the pipe-writer, then copy the in-memory buffer to the writer.
+/// <para/> When writing/reading the data-segment, we use an the <see cref="ObjectWriter"/>/<see cref="ObjectReader"/>
+/// subsystem.  This will write its own validation bits, and then the data describing the asset.  This data is:
+/// <code>
+/// ----------------------------------------------------------------------------------------------------------
+/// | data (variable length)                                                                                 |
+/// ----------------------------------------------------------------------------------------------------------
+/// | ObjectWriter validation (2 bytes) | checksum (16 bytes) | kind (1 byte) | asset-data (asset specified) |
+/// ----------------------------------------------------------------------------------------------------------
+/// </code>
+/// The validation bytes are followed by the checksum.  The checksum is needed in the message as assets can be found in
+/// any order (they are not reported in the order of the array of checksums passed into the writing method). Following
+/// this is the kind of the asset.  This kind is used by the reading code to know which asset-deserialization routine to
+/// invoke. Finally, the asset data itself is written out.
+/// </summary>
+internal readonly struct RemoteHostAssetWriter(
+    PipeWriter pipeWriter, Scope scope, AssetPath assetPath, ReadOnlyMemory<Checksum> checksums, ISerializerService serializer)
+{
+    /// <summary>
+    /// A sentinel byte we place between messages.  Ensures we can detect when something has gone wrong as soon as
+    /// possible. Note: the value we pick is neither ascii nor extended ascii.  So it's very unlikely to appear
+    /// accidentally.
+    /// </summary>
+    public const byte MessageSentinelByte = 0b10010000;
+
+    private static readonly ObjectPool<ReadWriteStream> s_streamPool = new(() => new());
+
+    private static readonly UnboundedChannelOptions s_channelOptions = new()
+    {
+        // We have a single task reading the data from the channel and writing it to the pipe.  This option allows the
+        // channel to operate in a more efficient manner knowing it won't have to synchronize data for multiple readers.
+        SingleReader = true,
+
+        // Currently we only have a single writer writing to the channel when we call FindAllAssetsAsync. However, we
```

Comment:
>FindAllAssetsAsync [](http://example.com/codeflow?start=86&length=18)

Maybe add a <see tag as the name might have changed.

Reply:
it's not a doc comment.  so a see-tag doesn't work. :(

Reply:
Wasn't FindAllAssetsAsync renamed?

Reply:
fixed.